lib/Semantic_check.ml

```diff
@@ -497,23 +250,182 @@ and inferred_unsizedtype_of_indexed loc ut typed_indexl =
 
-and semantic_check_expression cf ({expr; emeta} : untyped_expression) =
+(* == SEMANTIC CHECK OF PROGRAM ELEMENTS ==================================== *)
+
+(* Probably nothing to do here *)
+let semantic_check_assignmentoperator op = op
+
+(* Probably nothing to do here *)
+let semantic_check_autodifftype at = at
+
+(* Probably nothing to do here *)
+let rec semantic_check_unsizedtype = function
+  | Mir.UArray ut -> Mir.UArray (semantic_check_unsizedtype ut)
+  | UFun (l, rt) ->
+      UFun
+        ( List.map
+            ~f:(fun (at, ut) ->
+              (semantic_check_autodifftype at, semantic_check_unsizedtype ut)
+              )
+            l
+        , semantic_check_returntype rt )
+  | ut -> ut
+
+and semantic_check_returntype = function
+  | Mir.Void -> Mir.Void
+  | ReturnType ut -> ReturnType (semantic_check_unsizedtype ut)
+
+let semantic_error_e ({emeta; _} : Ast.typed_expression) msg =
+  semantic_error ~loc:emeta.loc msg
+
+(* -- Indentifiers ---------------------------------------------------------- *)
+let reserved_keywords =
+  [ "true"; "false"; "repeat"; "until"; "then"; "var"; "fvar"; "STAN_MAJOR"
+  ; "STAN_MINOR"; "STAN_PATCH"; "STAN_MATH_MAJOR"; "STAN_MATH_MINOR"
+  ; "STAN_MATH_PATCH"; "alignas"; "alignof"; "and"; "and_eq"; "asm"; "auto"
+  ; "bitand"; "bitor"; "bool"; "break"; "case"; "catch"; "char"; "char16_t"
+  ; "char32_t"; "class"; "compl"; "const"; "constexpr"; "const_cast"
+  ; "continue"; "decltype"; "default"; "delete"; "do"; "double"; "dynamic_cast"
+  ; "else"; "enum"; "explicit"; "export"; "extern"; "false"; "float"; "for"
+  ; "friend"; "goto"; "if"; "inline"; "int"; "long"; "mutable"; "namespace"
+  ; "new"; "noexcept"; "not"; "not_eq"; "nullptr"; "operator"; "or"; "or_eq"
+  ; "private"; "protected"; "public"; "register"; "reinterpret_cast"; "return"
+  ; "short"; "signed"; "sizeof"; "static"; "static_assert"; "static_cast"
+  ; "struct"; "switch"; "template"; "this"; "thread_local"; "throw"; "true"
+  ; "try"; "typedef"; "typeid"; "typename"; "union"; "unsigned"; "using"
+  ; "virtual"; "void"; "volatile"; "wchar_t"; "while"; "xor"; "xor_eq" ]
+
+let semantic_check_identifier id =
+  if id.name = !model_name then
+    Errors.semantic_error ~loc:id.id_loc
+      ("Identifier " ^ ("'" ^ id.name ^ "'") ^ " clashes with model name.") ;
+  if
+    String.is_suffix id.name ~suffix:"__"
+    || List.exists ~f:(fun str -> str = id.name) reserved_keywords
+  then
+    semantic_error ~loc:id.id_loc
+      ( "Identifier "
+      ^ ("'" ^ id.name ^ "'")
+      ^ " clashes with reserved keyword." ) ;
+  id
+
+(* -- Operators ------------------------------------------------------------- *)
+let semantic_check_operator i = i
+
+(* -- Expressions ----------------------------------------------------------- *)
+
+(* Function application validation checks *)
+let semantic_check_fn_map_rect ~loc id es =
+  match (id.name, es) with
+  | "map_rect", {expr= Variable arg1_name; _} :: _ ->
+      if
```

Comment:
Can this be written with a `when` clause more succinctly? 
```ocaml
let semantic_check_fn_map_rect ~loc id es =
  match (id.name, es) with
  | "map_rect", {expr= Variable arg1_name; _} :: _ 
      when
         String.(
          is_suffix arg1_name.name ~suffix:"_lp"
          || is_suffix arg1_name.name ~suffix:"_rng")
     -> Semantic_errors.FnMapRect (loc, arg1_name.name) |> Validate.error
  | _ -> Validate.ok ()
```

Reply:
Yes it can! I'll update